File: go/difference-of-squares/difference_of_squares.go
package diffsquares

// SquareOfSum calculates the square of the sum of natural numbers until the given limit.
func SquareOfSum(limit int) int {
	sum := 0
	for i := 1; i <= limit; i++ {
		sum = sum + i
	}
	return sum * sum
}

// SumOfSquares calculates the sum of the squares of natural numbers until the given limit.
func SumOfSquares(limit int) int {
	sum := 0
	for i := 1; i <= limit; i++ {
		sum = sum + i*i
	}
	return sum
}

// Difference calculates the difference between the square of the sum and the sum of the squares
// of natural numbers until the given limit.
func Difference(limit int) int {
	return SquareOfSum(limit) - SumOfSquares(limit)
}

File: ruby/difference-of-squares/difference_of_squares.rb
class Squares
  attr_reader :square_of_sum, :sum_of_squares, :difference

  def initialize(n)
    @square_of_sum = (1..n).sum**2
    @sum_of_squares = (1..n).map { |i| i**2 }.sum
    @difference = square_of_sum - sum_of_squares
  end
end
